content/blog/2022/2022-05-27-one-stop-guide-mapstruct.md

```diff
@@ -0,0 +1,2116 @@
+    ...
+}
+```
+
+The implementation class after generation would look something like below:
+
+```java
+@Generated(
+    value = "org.mapstruct.ap.MappingProcessor"
+)
+public class UserMapperImpl implements UserMapper {
+
+    private final ManagerMapper managerMapper = Mappers.getMapper(
+        ManagerMapper.class );
+    private final Validator validator = new Validator();
+
+    @Override
+    public BasicUserDTO convert(BasicUser user) throws ValidationException {
+        if ( user == null ) {
+            return null;
+        }
+
+        BasicUserDTOBuilder basicUserDTO = BasicUserDTO.builder();
+
+        basicUserDTO.id( validator.validateId( user.getId() ) );
+        basicUserDTO.name( user.getName() );
+
+        return basicUserDTO.build();
+    }
+
+    @Override
+    public BasicUser convert(BasicUserDTO userDTO) throws ValidationException {
+        if ( userDTO == null ) {
+            return null;
+        }
+
+        BasicUserBuilder basicUser = BasicUser.builder();
+
+        basicUser.id( validator.validateId( userDTO.getId() ) );
+        basicUser.name( userDTO.getName() );
+
+        return basicUser.build();
+    }
+    
+    ...
+}
+```
+
+MapStruct has automatically set the id of the mapper objects with the result of the `Validator` instance. It has added a `throws` clause for the method as well.
+
+## Mapper Retrieval Strategies
+
+To execute and call the mapper methods, we need to instantiate the mapper instance or the constructor. MapStruct provides various strategies to instantiate and access the generated mappers. Let’s look into each of them.
+
+### Mappers Factory
+
+If we are not using MapStruct as a Dependency Injection framework, then the mapper instances can be retrieved using the `Mappers` class. We need to invoke the `getMappers()` method from the factory passing the interface type of the mapper:
+
+```java
+UserMapper INSTANCE = Mappers.getMapper(UserMapper.class);
+```
+
+This pattern is one of the simplest ways to access the mapper methods. It can be accessed in the following way:
+
+```java
+PersonDTO personDTO = UserMapper.INSTANCE.convert(user,
+                                                  education,
+                                                  address,
+                                                  employment);
+```
+
+One thing to note is that the mappers generated by MapStruct are stateless and thread-safe. Thus it can be safely retrieved from several threads at the same time.
+
+### Dependency Injection
+
+If we want to use MapStruct in a dependency injection framework, then we need to access the mapper objects via dependency injection strategies and not use the `Mappers` class. MapStruct supports the component model for *CDI*(Contexts and Dependency Injection for Java EE) and the *Spring framework*.
+
+Let’s update our `UserMapper` class to work with Spring:
+
+```java
+@Mapper(componentModel = "spring", uses = {ManagerMapper.class, Validator.class})
```

Comment:
The `uses` attribute is taking the focus off the `componentModel` attribute. I suggest to remove the `uses` attribute so the reader knows what to focus on.

Reply:
Removed `uses` attribute.